core/trino-main/src/main/java/io/trino/sql/planner/LogicalPlanner.java

```diff
@@ -394,24 +395,48 @@ private RelationPlan createTableCreationPlan(Analysis analysis, Query query)
         }
 
         ConnectorTableMetadata tableMetadata = create.getMetadata().orElseThrow();
-
         Optional<TableLayout> newTableLayout = create.getLayout();
 
-        List<String> columnNames = tableMetadata.getColumns().stream()
-                .filter(column -> !column.isHidden()) // todo this filter is redundant
-                .map(ColumnMetadata::getName)
-                .collect(toImmutableList());
+        List<Symbol> visibleFieldMappings = getVisibleFieldMappings(plan);
 
         String catalogName = destination.getCatalogName();
         CatalogHandle catalogHandle = metadata.getCatalogHandle(session, catalogName)
                 .orElseThrow(() -> semanticException(CATALOG_NOT_FOUND, query, "Destination catalog '%s' does not exist", catalogName));
+
+        Assignments.Builder assignments = Assignments.builder();
+        ImmutableList.Builder<ColumnMetadata> finalColumnsBuilder = ImmutableList.builder();
+
+        forEachPair(tableMetadata.getColumns().stream(), visibleFieldMappings.stream(), (column, fieldMapping) -> {
```

Comment:
`forEachPair` is documented as being equivalent to zipping and zip notes:

> The resulting stream will only be as long as the shorter of the two input streams; if one stream is longer, its extra elements will be ignored.

This is IMO error-prone, unless there is a check that these two lists have equal length.

Reply:
This has been checked before in `MetadataManager`:

`checkState(newTableMetadata.getColumns().size() == columns.size(), "Table and query column count doesn't match");`

Reply:
This is actually not anymore valid because now the SPI is coercing individual columns instead of providing a full layout.